### PR TITLE
Tile coverage updates

### PR DESCRIPTION
**Pull request by brunoabinader**
This PR:
- Limits the amount of covered tiles for TileCoordinate-based coverage: we are now clamping the coverage to up to 4 (value is arbitrary) neighbor tiles in each direction
- Treats pan (prefetch) tiles as ideal, so they can be used to render the "horizon" in higher pitch values
- Increases the maximum pitch value to 67.5 degrees

screenshot from 2018-07-04 20-52-34

From the example above, the topmost tile is a higher zoom tile (configurable via `Map.prefetchZoomDelta()` that covers the remaining visible area.

Using pan tiles to cover the horizon reduces the amount of rendered tiles, and because they are barely visible, the lower fidelity doesn't account much. This PR is an ongoing effort to eventually support modifying the projection matrix eye according to the edge insets (https://github.com/mapbox/mapbox-gl-native/issues/12107) - modifying the eye position can also cause more tiles to be visible.

Most rendering failures in render tests are caused by setting a pitch value higher than maximum pitch, which causes different results between GL native (max pitch: 67.5) and GL JS (max pitch: 60.0).

/cc @asheemmamoowala @kkaefer @tmpsantos

### CODE REVIEW

**Comment by mb12** [comment]
@brunoabinader Can you please kindly clarify the following?

1. With this change does it mean that for pitch > maxpitch  irrespective of the width and height of the viewport, only 64 tiles at the most will be loaded?

2.  Can you please clarify why do tl, tr etc have to be rotated for the specific case of >maxpitch but not otherwise? Also why does the transformation include only the angle of rotation but it does not include the pitch? Also why is the rotation w.r.t. to (0, 0) and not the center of the map?

**Comment by brunoabinader** [comment]
Worth mentioning that using an arbitrary fixed limiting factor of 4 is a simpler solution to a somewhat complex problem. When implementing this, I've also considered the following:
- Calculate different limiting factors for `x` and `y`, taking width and height ratio in consideration.
- Take pitch in consideration when calculating the limiting factor (example below)
```C++
double rangeX = (topRight.x - topLeft.x) * std::cos(pitch);
```

For future consideration, we can also think of:
- Instead of two fixed tile layers (ideal and pan tileS), apply instead a mipmap-like approach that dynamically decreases the tile zoom as it is more distant from the center.

**Comment by brunoabinader** [comment]
> @brunoabinader In playing with this branch I see that the order in which tiles are fetched and rendered is prioritizing the panTiles over the idealTiles. This works great for satellite imagery. With vector tiles, Im sometimes seeing that low zoom tiles at the top edge of the screen are being loaded before tiles in the center or bottom of the screen. It's rare, but wondering if there's something to be done about it ?

I believe that's intended - my understanding of _pan tiles_ is that they prevent empty areas when panning - so we want to start showing off these faster than the higher zoomed tiles. However, we could add some logic to prevent this prioritization when _not_ gesturing (`TransformState` knows about that). Let's try out this logic and observe behavior.

> Is there a reason for enabling the clamping only past 60 degrees pitch ? It seems like it would be useful starting ~ 45 degrees.

We can modify that :+1: However, clamping past 60 degrees preserves the current behavior (because we've been limiting pitch to 60 degrees until now). 

Also, the amount of visible tiles also depends whether your zoom level is close to its integer value (e.g. `zoom 10.1` = less tiles), or closer its ceiling value (`zoom 10.95` = more tiles).

**Comment by brunoabinader** [comment]
Render tests `regressions/mapbox-gl-js#5911` and `regressions/mapbox-gl-js#5911a` are now failing when setting the tilt threshold for limiting the tile coverage to 45 instead of 60 degrees - this is actually one of the reasons why I selected 60 degrees as minimum threshold beforehand :sweat_smile:

**Comment by brunoabinader** [comment]
> This looks good to me. Any chance you can add a render test too

Good idea, I'll add one 👍

**Comment by brunoabinader** [comment]
@asheemmamoowala could you please re-review?

File: src/mbgl/util/tile_cover.cpp [hunk 1]
```diff
@@ -93,22 +97,51 @@ std::vector<UnwrappedTileID> tileCover(const Point<double>& tl,
...
+    if (mode == TileCoverMode::Limited5x5) {
+        limitTileCoordinate(tl);
+        limitTileCoordinate(tr);
+        limitTileCoordinate(br);
+        limitTileCoordinate(bl);
+    }
+
     auto scanLine = [&](int32_t x0, int32_t x1, int32_t y) {
-        int32_t x;
-        if (y >= 0 && y <= tiles) {
```
Comment (asheemmamoowala) [inline src/mbgl/util/tile_cover.cpp]: This condition is no longer necessary? I think it is necessary at the poles.

**Pull request closed without merging**